### PR TITLE
STORM-3541 report v2 metrics api in metrics tick

### PR DESCRIPTION
**Pull request by agresch**
We would like to be able to report v2 metrics api through the metrics tick, like v1 metrics.  

The main reason (besides keeping one reporter) is to reduce connection load on our openTSDB monitoring service.  Having one node doing reporting for a topology decreases the number of connections.

This also allows an easy migration path for users to the new v2 API.

### CODE REVIEW

File: docs/metrics_v2.md [hunk 1]
```diff
@@ -145,3 +145,6 @@ public interface StormMetricsFilter extends MetricFilter {
 }
 ```
 
+V2 metrics can also be reported to the Metrics Consumers registered with topology.metrics.consumer.register by enabling the 
+topology.enable.v2.metrics.tick configuration.
```
Comment (Ethanlm) [inline docs/metrics_v2.md]: nit: no newline needed

File: storm-client/src/jvm/org/apache/storm/daemon/Task.java [hunk 2]
```diff
@@ -252,6 +252,8 @@ private TopologyContext mkTopologyContext(StormTopology topology) throws IOExcep
             executor.getIntervalToTaskToMetricToRegistry(),
             executor.getOpenOrPrepareWasCalled(),
             workerData.getMetricRegistry());
+        executor.setMetricRegistry(workerData.getMetricRegistry());
```
Comment (Ethanlm) [inline storm-client/src/jvm/org/apache/storm/daemon/Task.java]: Should we put this line of code outside this function? They don't seem to be related

File: storm-client/src/jvm/org/apache/storm/executor/Executor.java [hunk 3]
```diff
@@ -310,17 +317,97 @@ public void metricsTick(Task task, TupleImpl tuple) {
...
+            return;
+        }
+        for (Map.Entry<String, Gauge> entry : stormMetricRegistry.registry().getGauges().entrySet()) {
+            String name = entry.getKey();
+            Object v = entry.getValue().getValue();
+            if (v instanceof Number) {
+                IMetricsConsumer.DataPoint dataPoint = new IMetricsConsumer.DataPoint(name, v);
+                dataPoints.add(dataPoint);
+            } else {
+                LOG.warn("Cannot report {}, it's value is not a Number {}", name, v);
```
Comment (Ethanlm) [inline storm-client/src/jvm/org/apache/storm/executor/Executor.java]: nit: `its`

File: storm-client/src/jvm/org/apache/storm/Config.java [hunk 4]
```diff
@@ -263,6 +263,13 @@
      */
     @IsPositiveNumber(includeZero = true)
     public static final String TOPOLOGY_METRICS_CONSUMER_CPU_PCORE_PERCENT = "topology.metrics.consumer.cpu.pcore.percent";
+
+    /**
+     * This config allows a topology to report metrics data points from the V2 metrics API through the metrics tick.
+     */
+    @IsBoolean
+    public static final String TOPOLOGY_ENABLE_V2_METRICS_TICK = "topology.enable.v2.metrics.tick";
```
Comment (Ethanlm) [inline storm-client/src/jvm/org/apache/storm/Config.java]: If this is enabled and `storm.metrics.reporters` is configured, will the metrics be sent to multiple places?

File: storm-client/src/jvm/org/apache/storm/daemon/Task.java (hunk 2 above)
Comment (Ethanlm) [reply storm-client/src/jvm/org/apache/storm/daemon/Task.java]: This should happen on Executor level instead of Task level.

File: docs/metrics_v2.md (hunk 1 above)
Comment (agresch) [reply docs/metrics_v2.md]: I'm not clear what newline you're referring to.  The one at the end of line 149?  Let me know and I will put up a PR with the other changes.

File: storm-client/src/jvm/org/apache/storm/Config.java (hunk 4 above)
Comment (agresch) [reply storm-client/src/jvm/org/apache/storm/Config.java]: Yes, v2 metrics would be reported by the V2 daemon reporter and the Topology v1 logging metrics consumer.

File: storm-client/src/jvm/org/apache/storm/daemon/Task.java (hunk 2 above)
Comment (agresch) [reply storm-client/src/jvm/org/apache/storm/daemon/Task.java]: Yes, I agree.  Will move.

File: docs/metrics_v2.md (hunk 1 above)
Comment (Ethanlm) [reply docs/metrics_v2.md]: The one at the end of line 148

**Comment by agresch** [comment]
@Ethanlm - updated
